### ISSUE TITLE
Testing story

### ISSUE THREAD

**Issue by jayphelps** [redux-observable/redux-observable#108]
We need to improve the documentation around testing. Partially the problem is that testing RxJS v5  isn't very well documented either, but we can't use that as an excuse forever.

If anyone feels they have figured out the ideal way to test epics, please share your thoughts here (or PR them too 😉)

**Comment by jayphelps** [comment]
Experimenting internally with some changes to RxJS + a `expectEpic` helper that allows this:

```js
const searchQueryEpic = (action$, store, api) =>
  action$.ofType(SEARCH_QUERY)
    .debounceTime(100)
    .switchMap(action =>
      api.search(action.payload)
        .map(searchQueryFulfilled)
    );

// test

expectEpic(searchQueryEpic, [
  'a----------b-c----------d-----------|',
  '----------a------------c----------d-|'
], {
  a: searchQuery('first'),
  b: searchQuery('second'),
  c: searchQuery('third'),
  d: searchQuery('fourth')
}, {
  a: searchQueryFulfilled('first_GOOD'),
  c: searchQueryFulfilled('third_GOOD'),
  d: searchQueryFulfilled('forth_GOOD')
}, null, {
  search: query => Observable.of(query + '_GOOD')
});
```

lmk your thoughts (if you can figure out what this is doing e.g. marbles and stuff)

**Comment by rgbkrk** [comment]
The marbles make sense here. The way I read this was that expectEpic takes

* epic
* marble diagram
* actions to hand to epic
* expected response
* `null` currently sits where I'd expect to pass the store
* `api` (which can be a mock)

**Comment by jayphelps** [comment]
@rgbkrk bingo.

**Comment by steida** [comment]
From a practical point of view, I would rather have the second arg reserved for any configurable dependencies, not a limited store only. A third argument is too much. I am using this pattern within injectMiddleware and it works well.

```js
// Like redux-thunk, but with just one argument.
const injectMiddleware = deps => ({ dispatch, getState }) => next => action =>
  next(typeof action === 'function'
    ? action({ ...deps, dispatch, getState }) // Note custom and default deps are mixed
    : action
  );
```

This is my current workaround
```js
import 'rxjs';
import * as appEpics from './app/epics';
import { combineEpics, createEpicMiddleware } from 'redux-observable';

const epics = [
  appEpics,
];

const configureEpics = (deps) => {
  // just extract functions from imported objects
  const flattenEpics = epics.reduce((prev, curr) =>
    prev.concat(Object.keys(curr).map(k => curr[k]))
  , []);
  // HACK Epic
  const epicsWithDeps = flattenEpics.map(epic =>
    (action$, store) => epic(action$, { ...store, ...deps })
  );
  const rootEpic = combineEpics(...epicsWithDeps);
  const epicMiddleware = createEpicMiddleware(rootEpic);
  return {
    epicMiddleware
  };
};

export default configureEpics;
```

So now I can use proper dependency injection.

```js
export const appStartEpic = (action$, { storageEngine }) => {
  console.log(storageEngine);
  return action$.ofType(APP_START)
    .delay(1000) // Asynchronously wait 1000ms then continue
    .mapTo({ type: 'PONG' });
}
```

Proposal for a new API

```js
  const epicMiddleware = createEpicMiddleware(rootEpic, deps); // That's all we need.
```

**Comment by jayphelps** [comment]
Overloading the store runs the risk of colliding with any methods redux adds to it later, and also prevents you from using any of the handy store mocking helpers (without awkward extra work)

**Comment by jayphelps** [comment]
@steida if you do prefer to merge the two, this is perfectly acceptable:

```js
const rootEpic = (action$, store) =>
  combineEpics(epic1, epic2)(action$, { ...deps, ...store });
```

**Comment by steida** [comment]
@jayphelps Aren't epic1, epic2 combined on every change? rootEpic is called only once?

**Comment by jayphelps** [comment]
@steida epics are basically factories, (typically) invoked only once, then actions pump through their Observables. I say typically, because there are ways of "spawning" short lived epics, but that isn't documented and not often used. So yep, rootEpic is called only once in a typical app.

**Comment by steida** [comment]
@jayphelps Perfect answer. Thank you. I'm in the middle of the process of adding redux-observables in https://github.com/este/este.

**Comment by jayphelps** [comment]
@steida 


This is how Rx works. You declaratively set up transforms using operators, then things are next'd into them.

```js
const rootEpic = (action$) => {
  console.log('only logs once');
  return new Observable(observer => {
    console.log('only logs once');
    return action$.subscribe(action => {
      console.log('logs for every action received');
    });
  });
};
```

**Comment by steida** [comment]
Still learning things. But I suppose redux-observable is the final answer for async redux.

**Comment by rgbkrk** [comment]
> I suppose redux-observable is the final answer for async redux

I welcome this.

**Comment by Casy** [comment]
what if i have a conditional, one to return an action, and other to return nothing? how can avoid the "undefined "type" property" error?

**Comment by jayphelps** [comment]
@Casy you'll need to provide some code for me to fully understand _where_ you're referring to, but you may be talking about something like this:

```js
const somethingEpic = action$
  action$.ofType(SOMETHING)
    .mergeMap(action => {
      if (action.payload === 'JUST....DOIT') {
        return Observable.ajaxOrWhatever()
          .map(response => somethingFulfilled(response));
      } else {
        // this 
        return Observable.empty();
      }
    });
```

Here, if the payload isn't what we want, we mergeMap an [`Observable.empty()`](http://reactivex.io/rxjs/class/es6/Observable.js~Observable.html#static-method-empty) which is an observable that emits nothing, it just complete()s. So when that condition happens, our epic won't emit anything either.

**Comment by Casy** [comment]
@jayphelps Observable.empty() is exactly what i'm looking for, once more thanks jey 👍

**Comment by cmelion** [comment]
I've been following the testing suggestions from the documentation and have been able to test most of my epics so far with some minor modifications:

```
 /**
 * expectEpic - Test helper for redux-observable epics
 *
 * @param epic     - the redux-observable epic to test
 * {
 * @param action   - the action the epic is listening for
 * @param call     - (optional) a local sinon.spy used to get/set/reset api calls
 * @param callArgs - (optional) An array containing the api method, followed by any params
 * @param done     - (optional) mocha callback for async tests
 * @param expected - marble notation for the expected result
 * @param replace  - (optional) allows replacement of fields using timestamps or other generated values
 * @param response - the expected payload
 * @param store    - (optional) a reference to the redux store
 * }
 */
 const expectEpic = (epic, {action, call, callArgs, done, expected, replace, response, store}) => {};
        
```

http://jsbin.com/dagiguw/edit?js,output

**Comment by jdetle** [comment]
Hi @jayphelps and others, I have think the way nteract tests rxjs and redux observable to be fairly low overhead, although maybe not for everyone.

```javascript

import { ActionsObservable } from 'redux-observable';
const chai = require('chai');
const expect = chai.expect;
const sinon = require('sinon');

import { executeCell } from '../../../src/notebook/actions';
import { executeCellEpic } from '../../../src/notebook/epics/execute';

describe('epic', () => {
  it('Epics with the appropriate input and output of actions', (done) => {
    const action$ = ActionsObservable.of(executeCell('id', 'source')));
    const actionBuffer = [];
    const responseActions = executeCellEpic(action$, store);
    const subscription = responseActions.subscribe(
      (x) => actionBuffer.push(x.payload.toString()), // Every action that goes through should get stuck on an array
      (err) => expect.fail(err, null), // It should not error in the stream
      () => {
        expect(actionBuffer).to.deep.equal(['Error: kernel not connected']); // ;
        done();
      },
    );
  });
```

is one test for the following epic:
```javascript
/**
 * the execute cell epic processes execute requests for all cells, creating
 * inner observable streams of the running execution responses
 */
export function executeCellEpic(action$, store) {
  return action$.ofType('EXECUTE_CELL')
    .do(action => {
      if (!action.id) {
        throw new Error('execute cell needs an id');
      }
      if (typeof action.source !== 'string') {
        throw new Error('execute cell needs source string');
      }
    })
    // Split stream by cell IDs
    .groupBy(action => action.id)
    // Work on each cell's stream
    .map(cellActionStream =>
      cellActionStream
        // When a new EXECUTE_CELL comes in with the current ID, we create a
        // a new stream and unsubscribe from the old one.
        .switchMap(({ source, id }) => createExecuteCellStream(action$, store, source, id))
    )
    // Bring back all the inner Observables into one stream
    .mergeAll()
    .catch(createErrorActionObservable(ERROR_EXECUTING));
}
```

**Comment by jayphelps** [comment]
@jdetle it shouldn't be necessary to provide an error handler mapping to `expect.fail`. If you don't provide a handler, it _should_ bubble and be caught as normal exceptions would and cause the test to fail with the correct formatting.

You might also find `.toArray()` a cleaner solution, but maybe not 😄 

```js
it('Epics with the appropriate input and output of actions', (done) => {
  const action$ = ActionsObservable.of(executeCell('id', 'source')));

  executeCellEpic(action$, store)
    .toArray()
    .subscribe(actions => {
      expect(actions).to.deep.equal([
        kernelNotConnected() // whatever action creators
      ]);

      done();
    });
});
```

**Comment by jdetle** [comment]
Ooh thanks, I'll have to check `.toArray` out. As an aside, I _really_ am a fan of your responses. So clear an concise :+1:.

**Comment by jayphelps** [comment]
@jdetle thanks! `toArray()` will basically drain the source, buffering each next'd item into an array until the source completes, once it completes, it then emits that single array of next'd values to the subscriber. So it handles that boilerplate for us. If the source never completes, the test will timeout, likely signalling that your epic is misbehaving (unless it's supposed to never complete naturally, which is very very rare)

**Comment by jayphelps** [comment]
@cmelion ultimately we're waiting for changes in RxJS itself, most notably TestScheduler needs to be autoinjected into any async operator e.g. `debounceTime` so that it can properly control virtual time. We also are working on a way to to represent arbitrary time durations; right now the marble diagrams display increments of 10 virtual ms per `-`, which isn't ideal in real world situations where you have large numbers like 1000ms and possibly non-divisible numbers like 555ms.

e.g. right now 1000ms represented in marble would have to be 100 dashes. `----------------------------------------------------------------------------------------------------` which is unmaintainable.

**Comment by cmelion** [comment]
@jayphelps yeah, the 100 dashes thing was a bit disconcerting for me.  I opted to pass the interval as part of the api to support mocking a different interval.

````js
export const api = {
    fetchIssues: () =>
        ajax.getJSON(BASE_URL),
    fetchInterval: () => 10000
};


const animateYoutrackIssuesEpic = (action$, store, call = indirect.call) =>
    action$.ofType(ActionTypes.ANIMATE_YOUTRACK_ISSUES, ActionTypes.PAUSE_ANIMATE_YOUTRACK_ISSUES)

        .switchMap(action => action.type === ActionTypes.ANIMATE_YOUTRACK_ISSUES ?

            /// If Animating then start the interval that cycles through issues and sets CSS animation state.
            Observable.interval(call(api.fetchInterval))
                .map(() => Actions.goToYoutrackIssue({
                    direction: 1,
                    count: action.count
                })) :

            /// If Paused then cancel the observable, stopping animation
            Observable.never()
        );

it('calls GOTO_YOUTRACK_ISSUE upon ANIMATE_YOUTRACK_ISSUES', (done) => {
    const payload = {issue: [{}, {}]};

    expectEpic(YouTrackIssuesEpic, {
        expected: ['-a', {
            a: {type: ActionTypes.GOTO_YOUTRACK_ISSUE, payload: {direction: 1, count: payload.issue.length}}
        }],
        action: ['a|', {
            a: {type: ActionTypes.ANIMATE_YOUTRACK_ISSUES, count: payload.issue.length}
        }],
        response: ['-'],
        call: call,
        callArgs: [api.fetchInterval],
        done: done
    });
});

      
````

**Comment by jyane** [comment]
I have found another test story.
IMO, The problem is native observable does not have `ofType` method.
`ofType` method is extended by ActionObservable.

So, I have extended `testScheduler.createColdeObservable` (add `ofType` method to cold).
Here is an example of test story.

``` js
const createCold = (testScheduler) => (marbles, values) => {
  const input$ = testScheduler.createColdObservable.bind(testScheduler)(marbles, values);

  /**
   * @see 'https://github.com/redux-observable/redux-observable/blob/fd393a1adf359381c98ca494bc5dc2cac80f6d07/src/ActionsObservable.js'
   */
  input$.ofType = (...keys) => {
    return input$.filter(({ type }) => {
      const len = keys.length;
      if (len === 1) {
        return type === keys[0];
      } else {
        for (let i = 0; i < len; i++) {
          if (keys[i] === type) {
            return true;
          }
        }
      }
      return false;
    });
  };

  return input$;
};
```

An epic will test
``` js
const fetchUser = (action$) =>
  action$.ofType('FETCH_USER').mergeMap((action) =>
    apiClient.getUser()
      .map((res) => res.data)
      .map(fetchUserSuccess)
      .catch((e) => Observable.of(fetchUserFailure(e)))
  );
```

Usage in Jest:
```js
describe('Test userEpic', () => {
  let testScheduler;
  let cold;

  beforeEach(() => {
    jest.resetModules();

    testScheduler = new TestScheduler((expected, actual) => {
      expect(expected).toEqual(actual);
    });

    cold = createCold(testScheduler);
  });

  it('should return FETCH_USER_SUCCESS when success', () => {
    jest.mock('src/utils/client', () => ({
      apiClient: {
        getUser: () => Rx.Observable.of({ data: { foo: 'bar' }})
      }
    }));
    const userEpic = require('src/epics/user').default;
    const input$  = cold('---a--', { a: { type: 'FETCH_USER' }});
    const expect$ =      '---b';
    const test$ = userEpic(input$, testScheduler);

    testScheduler.expectObservable(test$).toBe(expect$, { b: { type: 'FETCH_USER_SUCCESS', payload: { foo: 'bar' }}});
    testScheduler.flush();
  });
});

```

**Comment by jayphelps** [comment]
@jyane `ActionsObservable` expects a source observable as an argument, where it gets the actions from. So you can do this:

```js
const input$  = cold('---a--', { a: { type: 'FETCH_USER' }});
const action$ = new ActionsObservable(input$);
```

**Comment by jyane** [comment]
@jayphelps Thank you for your good suggestion! It is really useful for me.

**Comment by dfbaskin** [comment]
So given an epic that does something on a regular basis (using `interval`), I had created a test for it like this:

```javascript
import {ActionsObservable as Observable} from 'redux-observable'
import {VirtualTimeScheduler} from 'rxjs/scheduler/VirtualTimeScheduler';

import 'rxjs/add/observable/empty';
import 'rxjs/add/observable/interval';
import 'rxjs/add/operator/delay';
import 'rxjs/add/operator/concat';
import 'rxjs/add/operator/reduce';
import 'rxjs/add/operator/takeUntil';
import 'rxjs/add/operator/switchMap';
import 'rxjs/add/operator/toArray';
import 'rxjs/add/operator/do';

const intervalSeconds = 60;
let scheduler;

const intervalEpic = (action$) => {
    return Observable
        .interval(intervalSeconds * 1000, scheduler)
        .takeUntil(action$.reduce((v, action) => 0, 0))
        .switchMap(() => {
            return Observable.of({
                type: "SOME_ACTION"
            })
        });
};

export function runObservableEpicTest(observable, epic, done, verify) {
    let mockStore = {
        getState() {
            return {};
        }
    };
    Observable
        .empty()
        .concat(epic(observable, mockStore))
        .toArray()
        .do((actions) => verify(actions))
        .subscribe({
            error: (err) => done.fail(err),
            complete: () => done()
        });
}

describe('test with interval', () => {

    beforeEach(() => {
        scheduler = new VirtualTimeScheduler();
    });

    it('should allow testing of interval', (done) => {
        let observable = Observable
            .empty()
            .delay((intervalSeconds + 1) * 1000, scheduler);
        runObservableEpicTest(observable, intervalEpic, done, (actions) => {
            expect(actions).toEqual([
                { type: "SOME_ACTION" }
            ]);
        });
        scheduler.flush();
    });
});
```

But as you can see, I ended up having to pass the scheduler into the epic itself.  Not sure there is a way around that at this point (until perhaps the [testing updates](https://github.com/ReactiveX/rxjs/issues/1791) are released).

**Comment by hedgerh** [comment]
Regarding testing, testing is the only reason we went with redux-saga over redux-observable on our project.  We couldn't get the example test from the docs to work in the latest version of jest.  Nock and RxJS' ajax module didn't get along, and we couldn't mock out responses.  We tried to mock out ajax itself, but that quickly turned into a mess, as well.  We also tried figuring out how to wrap axios into an observable instead, but couldn't get it working.

Now, I know the ajax module and nock not working isn't `redux-observable`'s fault, but it's a pretty big obstacle to using `redux-observable`, and I think it should be addressed on your end, especially because you point to Redux's docs for setting up a test environment, which uses jest, as well.

Is this not something people are dealing with?  We tried really hard to get it working and absolutely could not.

**Comment by hedgerh** [comment]
Regarding testing, testing is the only reason we went with redux-saga over redux-observable on our project.  We couldn't get the example test from the docs to work in the latest version of jest.  Nock and RxJS' ajax module didn't get along, and we couldn't mock out responses.  We tried to mock out ajax itself, but that quickly turned into a mess, as well.  We also tried figuring out how to wrap axios into an observable instead, but couldn't get it working.

Now, I know the ajax module and nock not working isn't `redux-observable`'s fault, but it's a pretty big obstacle to using `redux-observable`, and we couldn't find any fixes for it.  I think it'd be a good idea to figure out what the problem is and properly document for people who would like to use observables in redux, especially since the redux-observable docs point to Redux's docs for setting up a testing environment, which uses jest.

**Comment by alyahmedaly** [comment]
any update on this

**Comment by jayphelps** [comment]
@alyahmedaly no

**Comment by alyahmedaly** [comment]
i used Injecting Dependencies and it's working as expected
`    const action$ = ActionsObservable.of(fromOrders.fetchOrders(1, 10));
    const dependencies = {
      orderService: {
        getOrders: () => Observable.of(response)
      }
    };

    const expectedAction = {
      type: fromOrders.actionTypes.FETCH_ORDERS_SUCCESS,
      page: response.orders.page,
      pageSize: response.orders.pageSize,
      totalItems: response.orders.totalItems,
      items: response.orders.items
    };

    return OrdersEpic(action$, store, dependencies)
      .toArray()
      .toPromise()
      .then(actions => expect(actions, [expectedAction]));`

**Comment by clivend** [comment]
I'm working on this now, maybe it helps
[https://github.com/redux-observable/redux-observable/issues/144#issuecomment-382648065](https://github.com/redux-observable/redux-observable/issues/144#issuecomment-382648065)

**Comment by huy-nguyen** [comment]
With the addition of `state$` stream in v1, I think marble testing is now feasible and is a great way to test epics. I wrote a [tutorial](https://www.huynguyen.io/marble-diagrams-redux-observable-2018-05/?utm_source=github&utm_medium=ref&utm_campaign=githubissue20180527) to show beginners how to do it. What do you think?
